src/bloom/bit-set.ts

```diff
@@ -0,0 +1,178 @@
+/* file : BitSet.ts
+MIT License
+
+Copyright (c) 2021 David Leppik
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+import {encode, decode} from "base64-arraybuffer";
+
+const bitsPerWord = 8;
+
+/**
+ * A memory-efficient Boolean array. Contains just the minimal operations needed for our Bloom filter implementation.
+ *
+ * @author David Leppik
+ */
+export default class BitSet {
+    readonly size: number;
+
+    // Uint32Array may be slightly faster due to memory alignment, but this avoids endianness when serializing
+    private array: Uint8Array;
+
+    /**
+     * Constructor. All bits are initially set to false.
+     * @param size the number of bits that can be stored. (This is NOT required to be a multiple of 8.)
+     */
+    constructor(size: number) {
+        this.size = size
+        this.array = new Uint8Array(Math.ceil(size / bitsPerWord))
+    }
+
+    /**
+     * Returns the value of the bit at the given index
+     * @param index position of the bit, zero-indexed
+     */
+    has(index: number): boolean {
+        const wordIndex = Math.floor(index / bitsPerWord)
+        const mask = 1 << (index % bitsPerWord)
+        return (this.array[wordIndex] & mask) !== 0
+    }
+
+    /**
+     * Set the bit to true
+     * @param index position of the bit, zero-indexed
+     */
+    add(index: number) {
+        const wordIndex = Math.floor(index / bitsPerWord)
+        const mask = 1 << (index % bitsPerWord)
+        this.array[wordIndex] = this.array[wordIndex] | mask
+    }
+
+    /**
+     * Set the bit to false
+     * @param index position of the bit, zero-indexed
+     */
+    remove(index: number) {
+        const wordIndex = Math.floor(index / bitsPerWord)
+        const mask = 1 << (index % bitsPerWord)
+        this.array[wordIndex] = this.array[wordIndex] ^ mask
+    }
```

Comment:
This function doesn't do what it says. It sets the bit to false only if it was set to 1 before, otherwise it switches it to 1.

Maybe you meant

 ```suggestion
    /**
     *  Set the bit to false if it is currently set to true, otherwise it set it to true
     * @param index position of the bit, zero-indexed
     */
    switch(index: number) {
        const wordIndex = Math.floor(index / bitsPerWord)
        const mask = 1 << (index % bitsPerWord)
        this.array[wordIndex] ^= mask
    }
```

or 

```suggestion
    /**
     * Set the bit to false
     * @param index position of the bit, zero-indexed
     */
    remove(index: number) {
        const wordIndex = Math.floor(index / bitsPerWord)
        const mask = ~0 ^ (1 << (index % bitsPerWord))
        this.array[wordIndex] &= mask
    }
```

Reply:
Good catch. I'm going to remove this method entirely, since it's not used and not properly tested. The class is only there to support the simple Bloom filter, which is write-only. If we don't use it, I don't want people to use it as sample code.